File: typescript-demo/src/CreateDatabaseRecordDemo.ts
/**
 * FlowUs API TypeScript Demo - 创建数据库记录演示
 * 基于JavaScript版本的CreateDatabaseRecordDemo转换而来
 */
import { FlowUsClient } from './utils/flowusClient';
import { config, validateConfig } from './config/apiConfig';
import createDatabaseDemo from "./CreateDatabaseDemo";

async function createDatabaseRecordDemo() {
  console.log('📄 FlowUs API TypeScript Demo - 创建数据库记录演示');
  console.log('================================================');
  
  try {
    // 验证配置
    validateConfig();
    
    // 创建API客户端
    const client = new FlowUsClient({
      baseURL: config.baseURL,
      apiToken: config.apiToken,
      timeout: config.timeout
    });
    // 临时创建一个新多维表用于测试记录添加
    const newDatabase = await createDatabaseDemo();
    
    // 构建页面数据 - 与JavaScript版本保持一致
    const pageData = {
      properties: {
        title: {
          type: 'title',
          title: [
            {
              type: 'text',
              text: {
                content: '复活的鲁路修V3'
              }
            }
          ]
        },
        '我的关联': {
          type: 'relation',
          relation: [
            {
              id: 'bcf00100-d7c9-4c5e-b755-7212bc34c26a'
            }
          ]
        },
        '网址': {
          type: 'url',
          url: 'google.com'
        },
        '人员': {
          type: 'people',
          people: [
            {
              object: 'user',
              id: 'e37e54a2-0278-48bd-a01f-ca3ed32f22bf'
            }
          ]
        },
        '已看': {
          type: 'number',
          number: 12
        },
        '我的电话': {
          type: 'phone_number',
          phone_number: '13683236049'
        },
        '评分': {
          type: 'number',
          number: '67'
        },
        '国家': {
          type: 'select',
          select: {
            name: '日本',
            color: 'yellow'
          }
        },
        '观看': {
          type: 'date',
          date: {
            start: '2025/07/03T15:16:00',
            end: null,
            time_zone: null
          }
        },
        '我的邮箱': {
          type: 'email',
          email: 'caokeji@gmail.com'
        },
        '完成': {
          type: 'checkbox',
          checkbox: true
        },
        '分类': {
          type: 'multi_select',
          multi_select: [
            {
              name: '动漫',
              color: 'grey'
            },
            {
              name: '魔幻',
              color: 'brown'
            }
          ]
        },
        '进度条': {
          type: 'number',
          number: '12'
        }
      },
      parent: {
        type: 'database_id',
        database_id: newDatabase.id
      },
      icon: {
        type: 'emoji',
        emoji: '😁'
      }
    };
    
    
    // 显示属性详情
    console.log('\n📝 页面属性详情:');
    Object.entries(pageData.properties).forEach(([key, prop]: [string, any]) => {
      let value = '';
      switch (prop.type) {
        case 'title':
          value = prop.title[0].text.content;
          break;
        case 'number':
          value = prop.number;
          break;
        case 'url':
          value = prop.url;
          break;
        case 'email':
          value = prop.email;
          break;
        case 'phone_number':
          value = prop.phone_number;
          break;
        case 'checkbox':
          value = prop.checkbox ? '已选中' : '未选中';
          break;
        case 'select':
          value = prop.select ? `${prop.select.name} (${prop.select.color})` : '无选择';
          break;
        case 'multi_select':
          value = prop.multi_select && prop.multi_select.length > 0 
            ? prop.multi_select.map((item: any) => `${item.name} (${item.color})`).join(', ')
            : '无选择';
          break;
        case 'date':
          value = prop.date.start;
          break;
        case 'people':
          value = prop.people && prop.people.length > 0 ? `${prop.people.length} 个用户` : '无用户';
          break;
        case 'relation':
          value = `${prop.relation.length} 个关联`;
          break;
        case 'files':
          value = prop.files && prop.files.length > 0 ? `${prop.files.length} 个文件` : '无文件';
          break;
        default:
          value = '复杂类型';
      }
      console.log(`  • ${key} (${prop.type}): ${value}`);
    });
    
    console.log('\n🚀 开始创建页面记录...');
    
    // 调用API创建页面
    const result = await client.createPage(pageData);
    
    console.log('✅ 页面记录创建成功！');
    console.log('页面ID:', result.id);
    console.log('页面URL:', result.url || '无');
    console.log('创建时间:', result.createdAt);
    console.log('最后更新时间:', result.updatedAt);
    
    // 显示创建者信息
    if (result.createdBy) {
      console.log('创建者ID:', result.createdBy.id);
    }
    
    // 显示父级信息
    if (result.parent) {
      console.log('父级类型:', result.parent.type);
      if (result.parent.page_id) {
        console.log('所在页面ID:', result.parent.page_id);
      } else if (result.parent.database_id) {
        console.log('所在数据库ID:', result.parent.database_id);
      }
    }
    
    // 显示图标信息
    if (result.icon) {
      console.log('页面图标:', result.icon.emoji || (result.icon.external?.url || '无'));
    }
    
    // 显示属性信息
    if (result.properties) {
      console.log('\n📋 已设置的属性:');
      Object.entries(result.properties).forEach(([key, prop]: [string, any]) => {
        console.log(`  • ${key}: ${prop.type}`);
      });
    }
    
    console.log('\n🎉 创建数据库记录演示完成！');
    console.log('💡 提示: 您可以在FlowUs中查看刚创建的页面记录');
    
    return result;
    
  } catch (error: any) {
    console.error('❌ 创建数据库记录演示失败:', error);
    if (error.response?.data) {
      console.error('详细错误信息:', JSON.stringify(error.response.data, null, 2));
    }
    
    // 提供具体的错误解决建议
    if (error.message.includes('404') || error.message.includes('not found')) {
      console.error('\n💡 可能的解决方案:');
      console.error('1. 检查数据库ID是否正确');
      console.error('2. 确保您有该数据库的写入权限');
      console.error('3. 在.env文件中配置正确的DEMO_DATABASE_ID');
    } else if (error.message.includes('400') || error.message.includes('validation')) {
      console.error('\n💡 可能的解决方案:');
      console.error('1. 检查属性值格式是否正确');
      console.error('2. 确保选择项的ID存在于数据库中');
      console.error('3. 检查日期格式是否正确');
    } else if (error.message.includes('403') || error.message.includes('forbidden')) {
      console.error('\n💡 可能的解决方案:');
      console.error('1. 检查API Token是否有效');
      console.error('2. 确保您有创建页面的权限');
    }
    
    process.exit(1);
  }
}

// 如果直接运行此文件，则执行演示
if (require.main === module) {
  createDatabaseRecordDemo();
}

export default createDatabaseRecordDemo; 

File: java-demo/src/main/java/cn/flowus/demo/CreateDatabaseRecordDemo.java
package cn.flowus.demo;
import cn.flowus.demo.config.ApiConfig;
import org.openapitools.client.ApiClient;
import org.openapitools.client.ApiException;
import org.openapitools.client.auth.*;
import org.openapitools.client.api.DefaultApi;
import org.openapitools.client.model.*;

import java.math.BigDecimal;
import java.net.URI;
import java.util.*;


public class CreateDatabaseRecordDemo {
    public static void main(String[] args) throws Exception{
        // 加载配置并初始化API客户端
        ApiConfig config = ApiConfig.getInstance();
        config.printConfig();
        
        // 检查配置是否有效
        if (!config.isConfigValid()) {
            System.err.println("❌ 配置无效，请检查 .env 文件中的 FLOWUS_BEARER_TOKEN 设置");
            System.err.println("💡 请复制 .env.example 为 .env 并填入您的实际配置");
            return;
        }
        
        ApiClient apiClient = config.getApiClient();
        DefaultApi apiInstance = new DefaultApi(apiClient);
        
        // 可以是从 CreateDatabaseDemo中创建的多维表
        String databaseId = "66455130-a811-4d30-8132-f4479e2ca7bb";
        
        // 创建页面请求对象
        CreatePageRequest createRecordRequest = new CreatePageRequest();
        
        // 创建属性映射
        Map<String, CreatePagePropertyValue> properties = new HashMap<>();
        
        // 添加标题属性
        CreatePagePropertyValueTitle titleProperty = new CreatePagePropertyValueTitle();
        titleProperty.setType("title");
        List<RichTextItem> titleText = new ArrayList<>();
        RichTextItem titleItem = new RichTextItem()
            .type(RichTextItem.TypeEnum.TEXT)
            .text(new RichTextItemText().content("复活的鲁路修V3"));
        titleText.add(titleItem);
        titleProperty.setTitle(titleText);
        properties.put("title", new CreatePagePropertyValue(titleProperty));
        
        // 添加关联属性
        CreatePagePropertyValueRelation relationProperty = new CreatePagePropertyValueRelation();
        relationProperty.setType("relation");
        List<CreatePagePropertyValueRelationRelationInner> relationList = new ArrayList<>();
        CreatePagePropertyValueRelationRelationInner relation = new CreatePagePropertyValueRelationRelationInner();
        relation.setId(UUID.fromString("bcf00100-d7c9-4c5e-b755-7212bc34c26a"));
        relationList.add(relation);
        relationProperty.setRelation(relationList);
        properties.put("我的关联", new CreatePagePropertyValue(relationProperty));
        
        // 添加网址属性
        CreatePagePropertyValueUrl urlProperty = new CreatePagePropertyValueUrl();
        urlProperty.setType("url");
        urlProperty.setUrl(URI.create("google.com"));
        properties.put("网址", new CreatePagePropertyValue(urlProperty));
        
        // 添加人员属性
        CreatePagePropertyValuePeople peopleProperty = new CreatePagePropertyValuePeople();
        peopleProperty.setType("people");
        List<CreatePagePropertyValuePeoplePeopleInner> peopleList = new ArrayList<>();
        CreatePagePropertyValuePeoplePeopleInner person = new CreatePagePropertyValuePeoplePeopleInner();
        person.setObject(CreatePagePropertyValuePeoplePeopleInner.ObjectEnum.USER);
        person.setId(UUID.fromString("e37e54a2-0278-48bd-a01f-ca3ed32f22bf"));
        peopleList.add(person);
        peopleProperty.setPeople(peopleList);
        properties.put("人员", new CreatePagePropertyValue(peopleProperty));
        
        // 添加已看属性（数字）
        CreatePagePropertyValueNumber numberProperty1 = new CreatePagePropertyValueNumber();
        numberProperty1.setType("number");
        numberProperty1.setNumber(new BigDecimal(12));
        properties.put("已看", new CreatePagePropertyValue(numberProperty1));
        
        // 添加电话属性
        CreatePagePropertyValuePhoneNumber phoneProperty = new CreatePagePropertyValuePhoneNumber();
        phoneProperty.setType("phone_number");
        phoneProperty.setPhoneNumber("13683236049");
        properties.put("我的电话", new CreatePagePropertyValue(phoneProperty));
        
        // 添加评分属性（数字）
        CreatePagePropertyValueNumber numberProperty2 = new CreatePagePropertyValueNumber();
        numberProperty2.setType("number");
        numberProperty2.setNumber(new BigDecimal(67));
        properties.put("评分", new CreatePagePropertyValue(numberProperty2));
        
        // 添加国家属性（选择）
        CreatePagePropertyValueSelect selectProperty = new CreatePagePropertyValueSelect();
        selectProperty.setType("select");
        CreatePagePropertyValueSelectSelect select = new CreatePagePropertyValueSelectSelect();
        select.setId("63214291-be18-403c-9e47-6da9e24e5d91");
        select.setName("日本");
        select.setColor("yellow");
        selectProperty.setSelect(select);
        properties.put("国家", new CreatePagePropertyValue(selectProperty));
        
        // 添加观看属性（日期）
        CreatePagePropertyValueDate dateProperty = new CreatePagePropertyValueDate();
        dateProperty.setType("date");
        CreatePagePropertyValueDateDate date = new CreatePagePropertyValueDateDate();
        date.setStart("2025/07/03T15:16:00");
        date.setEnd(null);
        date.setTimeZone(null);
        dateProperty.setDate(date);
        properties.put("观看", new CreatePagePropertyValue(dateProperty));
        
        // 添加邮箱属性
        CreatePagePropertyValueEmail emailProperty = new CreatePagePropertyValueEmail();
        emailProperty.setType("email");
        emailProperty.setEmail("caokeji@gmail.com");
        properties.put("我的邮箱", new CreatePagePropertyValue(emailProperty));
        
        // 添加完成属性（复选框）
        CreatePagePropertyValueCheckbox checkboxProperty = new CreatePagePropertyValueCheckbox();
        checkboxProperty.setType("checkbox");
        checkboxProperty.setCheckbox(false);
        properties.put("完成", new CreatePagePropertyValue(checkboxProperty));
        
        // 添加分类属性（多选）
        CreatePagePropertyValueMultiSelect multiSelectProperty = new CreatePagePropertyValueMultiSelect();
        multiSelectProperty.setType("multi_select");
        List<CreatePagePropertyValueSelectSelect> multiSelectList = new ArrayList<>();
        
        CreatePagePropertyValueSelectSelect option1 = new CreatePagePropertyValueSelectSelect();
        option1.setId("9d5ca623-36dc-4fbe-bd67-45d0df2c3dbc");
        option1.setName("动漫");
        option1.setColor("grey");
        multiSelectList.add(option1);
        
        CreatePagePropertyValueSelectSelect option2 = new CreatePagePropertyValueSelectSelect();
        option2.setId("c3f86e01-6a64-4bdd-8a09-db8e77f2d489");
        option2.setName("魔幻");
        option2.setColor("brown");
        multiSelectList.add(option2);
        
        multiSelectProperty.setMultiSelect(multiSelectList);
        properties.put("分类", new CreatePagePropertyValue(multiSelectProperty));
        
        // 添加进度条属性（数字）
        CreatePagePropertyValueNumber numberProperty3 = new CreatePagePropertyValueNumber();
        numberProperty3.setType("number");
        numberProperty3.setNumber(new BigDecimal(12));
        properties.put("进度条", new CreatePagePropertyValue(numberProperty3));
        
        // 设置属性
        createRecordRequest.setProperties(properties);
        
        // 设置父级（数据库ID）
        ParentDatabaseId parent = new ParentDatabaseId();
        parent.setType(ParentDatabaseId.TypeEnum.DATABASE_ID);
        parent.setDatabaseId(UUID.fromString(databaseId));
        createRecordRequest.setParent(new Parent(parent));
        
        // 设置图标
        IconEmoji icon = new IconEmoji();
        icon.setType(IconEmoji.TypeEnum.EMOJI);
        icon.setEmoji("😁");
        createRecordRequest.setIcon(new Icon(icon));

        try {
            CreatePageResponse result = apiInstance.createPage(createRecordRequest);
            System.out.println(result);
        } catch (ApiException e) {
            System.err.println("Exception when calling DefaultApi#createDatabase");
            System.err.println("Status code: " + e.getCode());
            System.err.println("Reason: " + e.getResponseBody());
            System.err.println("Response headers: " + e.getResponseHeaders());
            e.printStackTrace();
        }
    }
}
